doc/man3/EVP_PKEY_copy_parameters.pod

```diff
@@ -47,14 +47,17 @@ EVP_PKEY_copy_parameters() is to handle public keys in certificates where the
 parameters are sometimes omitted from a public key if they are inherited from
 the CA that signed it.
 
-Since OpenSSL private keys contain public key components too the function
-EVP_PKEY_eq() can also be used to determine if a private key matches
-a public key.
-
 The deprecated functions EVP_PKEY_cmp() and EVP_PKEY_cmp_parameters() differ in
-their return values  compared to other _cmp() functions. They are aliases for
+their return values compared to other _cmp() functions. They are aliases for
 EVP_PKEY_eq() and EVP_PKEY_parameters_eq().
 
+The function EVP_PKEY_cmp() previously only checked the key parameters
+(if there are any) and the public key, assuming that there always was
+a public key and that private key equality could be derived from that.
+This is no longer necessarily true, as the public key may not be present.
+EVP_PKEY_eq(), and therefore also EVP_PKEY_cmp(), now compare all
+components that are or should be present in both keys.
```

Comment:
This does not seem to match the rest of the documentation nor the implementation. What I understand from this is that if the public and private key are both available, both will be checked. The implementation and the change in provider-keymgmt.pod are happy with just checking the public key.

Reply:
The assumption is that if an EVP_PKEY contains both a public and private part, they match each other (internal consistency).  Therefore, it should be safe to assume that a comparison of the public part is implicitly a comparison of the private part as well and vice versa, even if that comparison isn't done explicitly.  This is, BTW, exactly the assumption that the 1.1.1 and older EVP_PKEY_cmp makes.

So, would you say that extending this documentation to explain that would be helpful?

Reply:
> So, would you say that extending this documentation to explain that would be helpful?

It would not hurt: FWIW, such explicit wording would have helped me understand things earlier.

Reply:
Okie.  I've tried to explain a bit further.

Reply:
@baentsch, perhaps there's one thing that's not come through clearly enough yet: the provider is ultimately the code that determines exactly how to compare two keys, specifically because the providers have the internal knowledge.  In that sense, libcrypto knows absolutely nothing, except that it holds pointers to two provider owned keys, and asks the provider to compare, with what little indication it knows how to give.